File: lab2/spellchecker/loosely-coupled/src/spellchecker.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "document.h"
#include "dictionary.h"


/* tamaño maximo de una palabra */
#define MAX_WORD_SIZE 30

/* diccionario principal */
Dictionary main_dict;
/* diccionario de palabras ignoradas */
Dictionary ignored;
/* Documento a procesar */
Document doc;


/*******************************************************************
 * NAME :            int is_known(char *word)
 *
 * DESCRIPTION :     Verifica si una palabra es "conocida", ya sea 
 *                   porque esta en el diccionario principal o en el 
 *                   diccionario de palabras ignoradas. 
 *
 * PARAMETERS:
 *      INPUT:
 *           char    *word       Palabra a verificar.
 *
 * RETURN :
 *           Type: int
 *           Values: 1 si la palabra es conocida
 *                   0 si la palabra no es conocida
 *******************************************************************/
int is_known(char *word){
    /* completar aca  */
    int a = 0,
        b = 0,
        res = 0;
    if (word != NULL) {
        a = dict_contains(main_dict, word);
        b = dict_contains(ignored, word);
    }
    res = (a == 1 || b == 1)? 1: 0;
    return res;
}

/*******************************************************************
 * NAME :            void consult_user(char *word)
 *
 * DESCRIPTION :     Consulta al usuario sobre que accion realizar 
 *                   (aceptar, ignorar o reemplazar) con la palabra w.
 *                   Una vez que el usuario elige, realiza la accion 
 *                   elegida.
 *
 * PARAMETERS:
 *      INPUT:
 *           char    *word       Palabra sobre la cual se consulta la 
 *                            accion a realizar.
 *
 * RETURN :
 *           Type: void
 *******************************************************************/
void consult_user(char *word){
    char ans[2];
    if (word != NULL) {
        do{
            printf("Palabra no reconocida: %s\n Aceptar (a) - Ignorar (i) - Reemplazar (r): ", word);
            scanf("%s", ans);
        }while((strcmp(ans,"r") != 0) && (strcmp(ans,"a") != 0) && (strcmp(ans,"i") != 0));
        /* completar aca  */
        switch(ans[0]) {
            case 'r':
                /* limpio word */
                for (int i=0; i<MAX_WORD_SIZE; i++) { word[i] = '\0'; }
                printf("Corregir a: ");
                scanf("%s", word);
                break;
            case 'a':
                dict_add(main_dict, word);
                break;
            case 'i':
                dict_add(ignored, word);
                break;
            default:
                printf("Opcion incorrecta.\n");
        }
    }
}

/*******************************************************************
 * NAME :            void process_document(char *fname)
 *
 * DESCRIPTION :     Procesa el documento fname, palabra por palabra, 
 *                   consultando al usuario sobre la accion a realizar 
 *                   si la palabra no es conocida.
 * PARAMETERS:
 *      INPUT:
 *           char    *fname   Nombre del archivo a procesar.
 *
 * RETURN :
 *           Type: void
 *******************************************************************/
void process_document(char *fname){
    char current_word[MAX_WORD_SIZE];
    /* completar aca */
    if (fname == NULL) {
        printf("Nombre del documento invalido.\n");
        exit(1);
    }
    /* Creo los documentos */
    doc = calloc(1, sizeof(Document)+10);
    if (doc != NULL) {
        /* Abro los documentos */
        doc_open(doc, fname, "out.txt");

        while(doc_get_word(doc, current_word) == 1) {
            if (is_known(current_word) == 0 && (int)strlen(current_word) > 0) {
                consult_user(current_word);
            }
            doc_put_word(doc, current_word);
            for(int i=0; i<MAX_WORD_SIZE; i++) { current_word[i] = '\0'; }
        }
        doc_close(doc);
        free(doc);
        doc = NULL;
    }
}

/*******************************************************************
 * NAME :            int main(int argc, char **argv)
 *
 * DESCRIPTION :     Punto de entrada principal. Abre el diccionario 
 *                   principal, procesa el archivo especificado y 
 *                   guarda los cambios realizados en el diccionario 
 *                   principal.
 *******************************************************************/
int main(int argc, char **argv){
    char *dict;
    /* Verificamos el nro de argumentos. */
    if (argc < 2)
    {
        printf("spellchecker.c: nro de argumentos erroneo. Deben ser <documento> [<diccionario>].\n");
        return (1);
    }
    /* si se especifico un diccionario lo usamos,  */
    /* caso contrario usamos el diccionario por defecto */
    dict = (argc >=3) ? argv[2] : "dict.txt";

    /* completar aca */
    main_dict = dict_new();
    ignored = dict_new();
    dict_load(main_dict, dict);
    process_document(argv[1]);
    dict_save(main_dict, dict);
    printf("El documento %s ha sido procesado. Resultados en out.txt\n", argv[1]);
    main_dict = dict_destroy(main_dict);
    ignored = dict_destroy(ignored);
    return 0;
}

File: lab5/spellchecker/src/Spellchecker.java
import java.util.Scanner;

import document.Document;
import dictionary.Dictionary;
import dictionary.FileDictionary;
import dictionary.MemDictionary;
import word.Word;

/**
 * Esta clase representa a Spellchecker.
 * Conforma el default package.
 * @version 0.1
 * @author Fernando Copa.
 * @author Ramiro Della Vedova.
 */

public class Spellchecker {

    /**
     * Constructor de la clase.
     */
    public Spellchecker() {}

    /**
     * Consulta al usuario sobre que accion realizar (aceptar, ignorar o reemplazar) con la palabra word.
     * Una vez que el usuario elige, realiza la accion elegida.
     * @param word    Palabra sobre la cual se consulta la accion a realizar.
     * @param main_dict    Diccionario principal.
     * @param main_dict    Diccionario de palabras ignoradas.
     * @return Word    Palabra ya consultada, o nueva palabra.
     * @throws IOException    Si se produce un error.
     */
    public static Word consultUser(Word word, Dictionary main_dict, Dictionary ignored) {
        try {
            char c;    // Caracter para la opcion de accion a realizar.
            Scanner reader_char = new Scanner(System.in).useDelimiter("\\s*");    // Objeto Scanner del caracter.
            do {
                reader_char.reset();    // Se limpia el scanner.
                System.out.println ("Palabra no reconocida: '"+word.getWord()+"'. Aceptar (a) - Ignorar (i) - Reemplazar (r): ");
                c = reader_char.next().charAt(0);    // Se lee el caracter.
                reader_char.reset();    // Se limpia el scanner.
            } while ((c != 'a') && (c != 'i') && (c != 'r'));
            switch (c) {
                case 'a':    // Agrega a diccionario principal.
                    main_dict.add(word);
                    break;
                case 'i':    // Agrega a diccionario de palabras ignoradas.
                    ignored.add(word);
                    break;
                case 'r':    // Reemplaza la palabra.
                    String new_word = "";    // Nueva palabra.
                    Scanner reader_word = new Scanner(System.in);    // Objeto scanner de la nueva palabra.
                    System.out.println ("Ingrese una nueva palabra: ");
                    new_word = reader_word.nextLine();    // Lee la palabra.
                    word.setWord(new_word);    // Pasa la palabra a Word.
                    break;
                default:
                    break;
            }
        } catch(Exception msg) {
            System.out.println("Error:" + msg.getMessage());
            System.exit(1);    // Salida forzosa del programa.
        };
        return word;
    }

    /**
     * Procesa el documento fname_in, palabra por palabra, consultando al usuario sobre la accion a realizar
     * si la palabra no es conocida.
     * @param fname_in    Nombre del archivo a procesar.
     * @param fname_out    Nombre del archivo de salida.
     * @param main_dict    Diccionario principal.
     * @param main_dict    Diccionario de palabras ignoradas.
     * @return void
     * @throws IOException    Si se produce un error.
     */
    public static void processDocument(String fname_in, String fname_out, Dictionary main_dict, Dictionary ignored) {
        try {
            Word current_word = new Word();    // Palabra a procesar.
            Document doc = new Document(fname_in, fname_out);    // Abre documento de entrada y salida.
            while (true) {
                current_word = doc.getWord();    // Lee una palabra del documento de entrada.
                if ((current_word.getWord()) == "") {    // Establece si hay mas palabras par leer.
                    break;
                }
                if (!(main_dict.contains(current_word)) && !(ignored.contains(current_word))) {    // ¿Es una palabra desconocida?
                    current_word = consultUser(current_word, main_dict, ignored);    // Palabra desconocida.
                }
                doc.putWord(current_word);    // Agrega una palabra al documento de salida.
            }
            doc.close();    // Cierra el documento de entrada y de salida.
        } catch(Exception msg) {
            System.out.println("Error:" + msg.getMessage());
            System.exit(1);    // Salida forzosa del programa.
        };
    }

    /**
     * Punto de entrada principal. Abre el diccionario principal, procesa el archivo especificado y
     * guarda los cambios realizados en el diccionario principal.
     */
    public static void main(String[] args) {
        String dict_path;    // Path del diccionario.
        int i = args.length;    // Argumentos de entrada.
        if (i < 1) {    // Verificamos el nro de argumentos.
            System.out.println ("spellchecker.java: nro de argumentos erroneo. Deben ser <documento> [<diccionario>].");
            System.exit(1);
        }
        dict_path = (i >= 2) ? args[1] : "dict.txt";    // No pasa un diccionario por defecto.
        FileDictionary main_dict = new FileDictionary();    // Se crea un diccionario principal.
        main_dict.load(dict_path);    // Se carga el diccionario principal.
        MemDictionary ignored = new MemDictionary();    // Se crea un diccionario de palabras ignoradas.
        processDocument(args[0], "out.txt", main_dict, ignored);    // Se procesa el documento.
        main_dict.save(dict_path);    // Se guarda el diccionario principal.
        main_dict.clear();    // Se destruye el diccionario principal.
        ignored.clear();    // Se destruye el diccionario de palabras ignoradas.
        System.out.println ("El documento '"+args[0]+ "' ha sido procesado. Resultados en out.txt");
    }
}
